### PR TITLE
Reland: fix(rocr/thunk): Revoke SVM-API GPU access on deregister, with register re-grant

### PR DESCRIPTION
**Pull request by ashetaia-amd**
## Motivation

Relands #6699 (reverted in 51c421c0 / #9187) with the fix for what broke it.

#6699 made deregister revoke GPU access with NO_ACCESS, but register never re-granted it. jemalloc reuses a VA without munmap, so that NO_ACCESS stays on the kernel SVM range and the GPU faults on the next register of the same address. Under ASan + XNACK the hip-tests hung on that fault and CI killed them on timeout, which is what turned ASan red.

## Technical Details

- Revoke SVM-API GPU access on deregister (original #6699, reapplied past the revert).
- Accept same-base registrations with differing spans (follow-on lost in the revert).
- kfdtest coverage for the same-page divergent-span case.
- Re-grant GPU access on SVM-API register (new). Register now sets ACCESS_IN_PLACE per GPU so a reused VA regains access. Has to be ACCESS_IN_PLACE, not ACCESS: userptr memory can't migrate, so ACCESS makes the fault handler try a migration that fails and still faults. Matches what fmm_map_mem_svm_api() already uses.
- kfdtest for the register re-grant path.

Complementary to Philip Yang's kernel patch (unmap svm range when all GPUs set to no-access); kernel unmaps on no-access, thunk revokes on deregister and re-grants on register. Both are needed.

## Issue Tracking

JIRA ID: ROCM-28520

## Test Plan

Built the full-ASan gfx942 stack on top of Philip's kernel patch and ran the failing hip-tests (MemoryTest1/2, GraphsTest2) with HSA_XNACK=1, comparing baseline, #6699, #6699 + this fix, and the ACCESS-instead-of-ACCESS_IN_PLACE variant.

## Test Result

- #6699 without this fix: MemoryTest1 hangs mid-suite on VM_L2_PROTECTION_FAULT.
- #6699 with this fix: completes all 376 cases, same as baseline (261 pass, 10 pre-existing fails).
- ACCESS variant: still hangs, so ACCESS_IN_PLACE is required.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

File: projects/rocr-runtime/libhsakmt/src/fmm.c [hunk 1]
```diff
@@ -1112,32 +1369,62 @@ static HSAKMT_STATUS fmm_register_mem_svm_api(HsaKFDContext *ctx,
...
-	args->attrs[1].value = HSA_SVM_FLAG_EXT_COHERENT;
-	pr_debug("Registering to SVM %p size: %ld\n", (void*)aligned_addr,
+	args->attrs[num_gpus + 1].value = HSA_SVM_FLAG_EXT_COHERENT;
+	/* Validate and reserve the tracking entry before touching kernel state,
+	 * so a same-base re-registration with a mismatched size is rejected
+	 * (INVALID_PARAMETER) without issuing a stray SET_ATTR - deregister is
+	 * keyed only by base address and could not disambiguate the two extents.
+	 * Identical (base, size) registrations are refcounted.
+	 */
+	ret = svm_api_range_get(fmm_ctx, (void *)aligned_addr, aligned_size);
```
Comment (cfreeamd) [inline projects/rocr-runtime/libhsakmt/src/fmm.c]: 🤖 AI-assisted pre-review (for @cfreeamd)

**[Optional]** This comment says a same-base re-registration with a *mismatched* size is rejected (`INVALID_PARAMETER`) and only identical `(base, size)` pairs are refcounted — but `svm_api_range_get()` does the opposite: it refcounts mismatched sizes together and grows the tracked extent to the largest span, and `SVMApiSharedPageTest` explicitly asserts the differing-size second registration must **not** be rejected. The comment looks left over from an earlier reject-design; please update it to match the grow-extent behavior.

Relatedly, `svm_api_range_get()` only ever returns `HSAKMT_STATUS_SUCCESS`, so `if (ret != HSAKMT_STATUS_SUCCESS) return ret;` is unreachable today — and if it ever does return an error, this `return ret;` leaks the `args` buffer malloc'd just above. `goto out;` here would stay leak-safe.

File: projects/rocr-runtime/libhsakmt/src/fmm.c [hunk 2]
```diff
@@ -1102,6 +1129,236 @@ static HsaMemFlags fmm_translate_ioc_to_hsa_flags(uint32_t ioc_flags)
...
+		 * different page-aligned spans (e.g. small calloc'd buffers
+		 * packed into one page, one of which straddles into the next
+		 * page). They are independent registrations, so refcount them
+		 * together and grow the tracked extent to the largest span seen
+		 * at this base; deregister then revokes the full extent once the
+		 * last owner is gone, and overlap with other bases is resolved by
+		 * the interval subtraction in svm_api_range_put.
+		 */
+		++r->refcount;
+		if (aligned_size > r->size)
```
Comment (cfreeamd) [inline projects/rocr-runtime/libhsakmt/src/fmm.c:1179]: 🤖 AI-assisted pre-review (for @cfreeamd)

**[Optional]** `svm_api_range_get()` grows `r->size` to the larger span *before* the `SET_ATTR` ioctl, but the register error path only calls `svm_api_range_put()` to drop the refcount — it never restores the previous (smaller) extent. If a second, larger same-base registration fails the ioctl, the surviving first registration keeps tracking the grown extent, so its eventual deregister issues `NO_ACCESS` over pages the failed registration never actually got access to. Revoking a range you own is harmless, but if those extra pages belong to another live registration this over-revokes a neighbor. Only triggers on ioctl failure, but worth guarding (remember/restore the prior size, or grow only after the ioctl succeeds).

File: projects/rocr-runtime/libhsakmt/tests/kfdtest/src/KFDSVMRangeReGrant_test.cpp [hunk 3]
```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (C) 2026 Advanced Micro Devices, Inc. All Rights Reserved.
```
Comment (cfreeamd) [inline projects/rocr-runtime/libhsakmt/tests/kfdtest/src/KFDSVMRangeReGrant_test.cpp]: 🤖 AI-assisted pre-review (for @cfreeamd)

**[Nit]** New file uses the legacy full MIT license block; convention for new files is the short SPDX header:
```cpp
// Copyright © Advanced Micro Devices, Inc., or its affiliates.
//
// SPDX-License-Identifier: MIT
```

**Review by cfreeamd — COMMENTED** [review]
🤖 AI-assisted pre-review (for @cfreeamd)

Reviewed the rocr-runtime changes: libhsakmt/src/fmm.c (SVM-API range tracking + register re-grant / deregister revoke) and the new kfdtest coverage (KFDSVMRangeReGrant_test.cpp + the refcount / overlap / shared-page phases in KFDSVMRangeTest.cpp). The register/deregister symmetry, refcounting, and interval-subtraction revoke logic look correct and are well covered by the new tests. Left non-blocking notes inline on fmm.c (a comment that no longer matches the implementation, and an error-path extent-rollback edge case) plus an SPDX header nit on the new test file. All changes are in rocr-runtime scope.

**Review by dayatsin-amd — APPROVED** [review]

**Review by kentrussell — APPROVED** [review]

**Review by cfreeamd — COMMENTED** [review]
🤖 AI-assisted pre-review (for @cfreeamd) — follow-up after author update

Both prior **[Optional]** items are resolved:

- **Over-revoke on register ioctl failure:** commit `70415729` now calls `svm_api_range_get()` only after the `SET_ATTR` ioctl succeeds, so a failed register no longer inserts/grows a tracked extent that never got GPU access. Matches the suggested fix.
- **Stale reject-design comment / dead-code error return:** the register path was rewritten; the misleading `INVALID_PARAMETER` comment is gone and the tracking-result now flows through the shared `out:` free path (no leak).

Only the **[Nit]** remains: `KFDSVMRangeReGrant_test.cpp` still uses the legacy full MIT license block rather than the short `SPDX-License-Identifier: MIT` header — non-blocking.

**Review by cfreeamd — COMMENTED** [review]
🤖 AI-assisted pre-review (for @cfreeamd) — follow-up after author update (2026-08-15)

Prior [Optional] findings resolved:
- **Finding #2** (register ioctl failure → over-revoke): Fixed in commit 70415729 — `svm_api_range_get()` is now called *after* the `SET_ATTR` ioctl succeeds, so a failed register never grows a tracked extent or inserts a tracking entry.
- **Finding #1** (stale reject-design comment + dead-code/leaky error return): Resolved — the register path was rewritten; stale comment and dead `if (ret != SUCCESS) return ret;` are gone, freed cleanly via `goto out` on all paths.

**Finding #3** (SPDX header on `KFDSVMRangeReGrant_test.cpp`): Still the legacy full MIT block, not the short SPDX form. Non-blocking [Nit]; does not gate approval.

Gating approvals present: dayatsin-amd (2026-07-28), kentrussell (2026-07-29). All substantive findings resolved — ready for @cfreeamd approval.

File: projects/rocr-runtime/libhsakmt/tests/kfdtest/src/KFDSVMRangeReGrant_test.cpp (hunk 3 above)
Comment (cfreeamd) [reply projects/rocr-runtime/libhsakmt/tests/kfdtest/src/KFDSVMRangeReGrant_test.cpp]: Please use the SPDX header
